File: SDK/c/examples/basic_usage.c
/* basic_usage.c - Demonstrates connecting, pinging, listing modules, and
 * sending a synchronous command. Mirrors examples/basic_usage.py. */
#include "csm_tcp_router_client.h"

#include <stdio.h>
#include <stdlib.h>

#define HOST "localhost"
#define PORT 30007

int main(void) {
    /* 1. Wait until the server is ready (optional). */
    printf("Waiting for server ... ");
    fflush(stdout);
    csm_result_t r = csm_client_wait_for_server(HOST, PORT, 30000, 500);
    if (r != CSM_OK) {
        printf("TIMEOUT - server did not start within 30s.\n");
        return 1;
    }
    printf("ready.\n");

    /* 2. Create + connect. */
    csm_client_t *c = csm_client_create();
    if (!c) { fprintf(stderr, "Out of memory\n"); return 1; }

    r = csm_client_connect(c, HOST, PORT, 5000);
    if (r != CSM_OK) {
        fprintf(stderr, "Connection failed: %s\n", csm_result_str(r));
        csm_client_destroy(c);
        return 1;
    }
    printf("Connected to %s:%d\n", HOST, PORT);

    /* 3. Ping. */
    double ms = 0;
    r = csm_client_ping(c, 2000, &ms);
    if (r == CSM_OK) printf("Ping OK  latency=%.1f ms\n", ms);
    else             printf("Ping failed: %s\n", csm_result_str(r));

    /* 4. List CSM modules. */
    char *modules = NULL;
    if (csm_client_list_modules(c, &modules, 5000) == CSM_OK) {
        printf("\nLoaded modules:\n%s\n", modules);
        csm_string_free(modules);
    }

    /* 5. Send a synchronous command (uncomment when wired to a real module).
     *
     *  csm_command_response_t resp = {0};
     *  if (csm_client_send_and_wait(c, "API: Read -@ DAQmx", 5000, &resp) == CSM_OK) {
     *      printf("Sync response: %s\n", (char *)resp.raw);
     *  }
     *  csm_command_response_dispose(&resp);
     */

    /* 6. Disconnect & clean up. */
    csm_client_disconnect(c);
    csm_client_destroy(c);
    printf("Disconnected.\n");
    return 0;
}

File: SDK/python/examples/basic_usage.py
"""Basic usage example for csm-tcp-router-client.

Prerequisites
-------------
A running CSM-TCP-Router server (LabVIEW app).  The reference app defaults
to port 30007.  Start it from ``CSM-TCP-Router(Server).vi``.

Install the SDK::

    pip install csm-tcp-router-client

Run this example::

    python basic_usage.py
"""


from csm_tcp_router import TcpRouterClient
from csm_tcp_router.exceptions import ConnectionError

HOST = "localhost"
PORT = 30007


def main() -> None:
    # -----------------------------------------------------------------------
    # 1. Wait until the server is ready (optional – useful during app startup)
    # -----------------------------------------------------------------------
    print("Waiting for server …", end=" ", flush=True)
    client = TcpRouterClient()
    ok = client.wait_for_server(HOST, PORT, timeout=30, retry_interval=0.5)
    if not ok:
        print("TIMEOUT – server did not start within 30 s.")
        return
    print("ready.")

    # -----------------------------------------------------------------------
    # 2. Connect (use as a context manager so disconnect is always called)
    # -----------------------------------------------------------------------
    with TcpRouterClient() as client:
        try:
            client.connect(HOST, PORT)
        except ConnectionError as exc:
            print(f"Connection failed: {exc}")
            return

        print(f"Connected to {HOST}:{PORT}")

        # -------------------------------------------------------------------
        # 3. Ping – verify round-trip latency
        # -------------------------------------------------------------------
        ok, ms = client.ping()
        if ok:
            print(f"Ping OK  latency={ms * 1000:.1f} ms")
        else:
            print("Ping failed.")

        # -------------------------------------------------------------------
        # 4. List CSM modules loaded on the server
        # -------------------------------------------------------------------
        modules = client.list_modules()
        print(f"\nLoaded modules:\n{modules}")

        # -------------------------------------------------------------------
        # 5. List the API for the first module (if any)
        # -------------------------------------------------------------------
        first_module = modules.strip().splitlines()[0] if modules.strip() else None
        if first_module:
            api_text = client.list_api(first_module)
            print(f"\nAPI for '{first_module}':\n{api_text}")

        # -------------------------------------------------------------------
        # 6. Send a synchronous command  (replace with a real API of yours)
        # -------------------------------------------------------------------
        # resp = client.send_and_wait("API: Read -@ DAQmx")
        # print(f"\nSync response: {resp.text}")

        # -------------------------------------------------------------------
        # 7. Send an asynchronous command  (server returns cmd-resp handshake)
        # -------------------------------------------------------------------
        # client.post("API: Start Sampling -> DAQmx")
        # print("Async command sent – waiting for async-resp …")
        # time.sleep(1)
        # if not client.async_response_queue.empty():
        #     ar = client.async_response_queue.get_nowait()
        #     print(f"Async-resp: {ar.text}")

        # -------------------------------------------------------------------
        # 8. Send a no-reply command
        # -------------------------------------------------------------------
        # client.post_no_reply("API: Reset ->| DAQmx")
        # print("No-reply command sent.")

        print("\nDone.")

    print("Disconnected.")


if __name__ == "__main__":
    main()
